Introduce MaybeCompareFn trait to allow for partial compare specializations

gatesn opened the pull request:
Currently an array needs to statically decide if it supports a compute function, prior to seeing the RHS or other arguments.

This change allows the array's compute function to see the RHS, and then decide it doesn't support the operation, enabling easier fallback.

This is only implemented for compare.

It has also highlighted the need for storage DType information to be available to an ExtScalar. Either by adding it to ExtDType, or adding it directly into that scalar variant. @a10y keen for your thoughts

Notes by @AdamGS:
To follow up @gatesn's work here, I added some other things to this PR:
- I cleared up the boolean compare implementation as they were basically the arrow-based fallback we already had
- Added an ALP `MaybeCompareFn` implementation that shows how the new interface looks

gatesn committed:
Maybe Compare

gatesn commented on vortex-array/src/array/constant/compute.rs:
```diff
@@ -99,30 +95,17 @@ impl SearchSortedFn for ConstantArray {
-            };
-
-            Ok(Array::from_arrow(&boolean_array, true))
+            None
```
We could centralize the Arrow Array<->Scalar compute here since the entry function will swap the arguments for us.

gatesn committed:
Maybe Compare

AdamGS commented on encodings/byte-bool/src/compute/mod.rs:
```diff
@@ -13,8 +13,8 @@ use vortex_scalar::Scalar;
-    fn compare(&self) -> Option<&dyn CompareFn> {
-        Some(self)
+    fn compare(&self, array: &Array, operator: Operator) -> Option<VortexResult<Array>> {
+        CompareFn::compare(self, array, operator)
```
nit - I like `other` here, `array` is very overloaded

gatesn replied:
Ah yeah, I had deleted this and had the IDE implement the trait. Maybe rename in the trait if you prefer it?

AdamGS reviewed (approved):
LGTM

AdamGS commented on vortex-array/src/array/varbin/compute/compare.rs:
```diff
@@ -0,0 +1,77 @@
+
+    match arrow_lhs.data_type() {
+        DataType::Binary => {
+            let arrow_rhs = cast(arrow_rhs.as_ref(), &DataType::Binary)?;
```
Why is the `cast` needed here?

gatesn replied:
I guess we can have a fallible downcast? Not clear that the types are checked anywhere else

AdamGS reviewed (commented):
LGTM

AdamGS commented:
If there are no objections, I can start building on top of this PR. I think the general approach is good and we can start focusing on combinations we can/want to specialize.

AdamGS committed:
.

AdamGS committed:
Fix StructArray::filter length calculation (#769)

there's no need to calculate states here, we can just use the length of
a filtered array

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
cleanup, get clippy to pass and some small refactoring

AdamGS committed:
Merge branch 'develop' into ngates/compare-fn

AdamGS committed:
Merge branch 'develop' into adamg/alp-specialized-compare

AdamGS committed:
.

AdamGS committed:
merge ALP work

a10y commented:
@gatesn IIRC we had talked a couple months back about adding the storage DType to ExtensionDType. It makes sense to me that an ExtensionDType is just a logical type with some associated metadata. This still preserves flexibility of the physical encoding. I'd be a fan of adding the storage type into the ExtensionDType.

gatesn commented:
The problem is, that means we would have to use e.g. int64 as the storage type for all date times, regardless of the unit.

In our Python prototype with had an IntWidthArray that allowed you to "reinterpret cast" a narrower int type into a wider one, e.g. you could zero-copy from Arrow an int32 array but pretend that it's int64 in Vortex.

It would be kind of weird to canonicalize back to Arrow by somehow introspecting this though...

gatesn commented:
Ah, I had already thought of this: https://github.com/spiraldb/vortex/blob/0c4d74eb7afa359775493dea8d8de77841e727f5/vortex-scalar/src/extension.rs#L9-L10

There's something appealing about not prescribing the storage type in the DType. If we add it into the ExtScalar (it is already part of the ExtArray), then we have enough information in the right places.

robert3005 commented:
I thought the storage type of the extdtype was the canonicalized storage type, and not required storage type?

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
Merge branch 'develop' into ngates/compare-fn

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
.

AdamGS committed:
Merge branch 'develop' into ngates/compare-fn

AdamGS committed:
fix tests

robert3005 commented on vortex-array/src/array/varbin/compute/compare.rs:
```diff
@@ -0,0 +1,77 @@
+    let constant = rhs.slice(0, 1)?.clone();
+    let arrow_rhs = constant.into_canonical()?.into_arrow()?;
+
+    match arrow_lhs.data_type() {
```
you could compare the View types here

robert3005 replied:
Oh but this is us being converted into arrow, ignore me. We should have a test though that this still works once we change canonicalization

robert3005 commented on vortex-array/src/array/primitive/compute/compare.rs:
```diff
@@ -1,24 +1,48 @@
-        Ok(BoolArray::try_new(match_mask, validity)?.into_array())
+impl MaybeCompareFn for PrimitiveArray {
+    fn maybe_compare(&self, other: &Array, operator: Operator) -> Option<VortexResult<Array>> {
+        let fn_impl = || -> VortexResult<Array> {
```
You immediately invoke this function - why make it a function?

AdamGS replied:
So I can have nice `?` for errors, but I'll just split it into a different function

AdamGS replied:
done

AdamGS committed:
.

robert3005 commented on encodings/alp/src/compute.rs:
```diff
@@ -67,3 +80,84 @@ impl SliceFn for ALPArray {
+
+            match pvalue {
+                Some(PValue::F32(f)) => {
+                    let encoded = f32::encode_single(f, self.exponents());
+                    match encoded {
+                        Ok(encoded) => {
+                            let s = ConstantArray::new(encoded, self.len());
+                            Some(compare(&self.encoded(), s.as_array_ref(), operator))
+                        }
+                        Err(exception) => {
+                            if let Some(patches) = self.patches().as_ref() {
+                                let s = ConstantArray::new(exception, self.len());
+                                Some(compare(patches, s.as_array_ref(), operator))
+                            } else {
+                                Some(Ok(BoolArray::from_vec(
+                                    vec![false; self.len()],
+                                    Validity::AllValid,
+                                )
+                                .into_array()))
+                            }
+                        }
+                    }
```
I think if you make a function generic over ALPFloat you can reduce the repetition

AdamGS replied:
good idea, done + fixed a bug

robert3005 commented on encodings/alp/src/compute.rs:
```diff
@@ -67,3 +80,84 @@ impl SliceFn for ALPArray {
+                    match encoded {
+                        Ok(encoded) => {
+                            let s = ConstantArray::new(encoded, self.len());
+                            Some(compare(&self.encoded(), s.as_array_ref(), operator))
```
I wonder if I can get rid of AsArray since we no longer have lifetimes

robert3005 replied:
anyway, just us AsRef or `&`

robert3005 commented on encodings/alp/src/compute.rs:
```diff
@@ -67,3 +80,84 @@ impl SliceFn for ALPArray {
+    }
+}
+
+impl MaybeCompareFn for ALPArray {
```
tests?

AdamGS replied:
added a bunch of basic ones

AdamGS committed:
.

AdamGS committed:
fn and tests

AdamGS committed:
.

robert3005 commented on encodings/alp/src/compute.rs:
```diff
@@ -161,3 +121,66 @@ impl MaybeCompareFn for ALPArray {
     }
 }
+
+fn alp_scalar_compare<F: ALPFloat + Into<Scalar>>(
```
you don't need `Into<Scalar>`?

AdamGS replied:
I do, to create the `ConstantArray` (for either values)

AdamGS committed:
Add test for alp-const-null compare

AdamGS committed:
.

robert3005 commented on vortex-array/src/array/varbin/compute/compare.rs:
```diff
@@ -63,3 +63,52 @@ fn compare_constant_arrow<T: ByteArrayType>(
+
+    #[test]
+    fn basic_test() {
+        let x = vec![
+            b"one".as_slice(),
+            b"two".as_slice(),
+            b"three".as_slice(),
+            b"four".as_slice(),
+            b"five".as_slice(),
+            b"six".as_slice(),
+        ]
+        .into_iter()
+        .flat_map(|x| x.iter().cloned())
+        .collect_vec();
+
+        let bytes = PrimitiveArray::from(x).to_array();
+        let offsets = PrimitiveArray::from(vec![0, 3, 6, 11, 15, 19, 22]).to_array();
+        let arr = VarBinArray::try_new(
+            offsets,
+            bytes,
+            DType::Utf8(Nullability::Nullable),
+            Validity::AllValid,
+        )
+        .unwrap();
```
There's a VarBinBuilder now

AdamGS committed:
Use VarBinBuilder

AdamGS committed:
Add default

AdamGS committed:
.
